sdk/src/Core/Amazon.Runtime/Credentials/DefaultInstanceProfileAWSCredentials.cs

```diff
@@ -75,97 +81,109 @@ public static DefaultInstanceProfileAWSCredentials Instance
+
             // if IMDS is turned off, no need to spin up the timer task
             if (!EC2InstanceMetadata.IsIMDSEnabled)
             {
                 return;
             }
 
-            _logger = Logger.GetLogger(typeof(DefaultInstanceProfileAWSCredentials));
+            _credentialsRetrieverTimer = new Timer(RenewCredentials, null, TimeSpan.Zero, _neverTimespan); // This invokes synchronous calls in seperate thread.
+            FeatureIdSources.Add(UserAgentFeatureId.CREDENTIALS_IMDS);
+        }
+
+        internal static DefaultInstanceProfileAWSCredentials CreateTestDefaultInstanceProfileAWSCredentials(
+            IIMDSAccessMethods imdsAccessMethods,
+            TimeSpan refreshRate)
+        {
+            return new DefaultInstanceProfileAWSCredentials(
+                imdsAccessMethods,
+                refreshRate);
+        }
+
+        private DefaultInstanceProfileAWSCredentials(
+            IIMDSAccessMethods imdsAccessMethods,
+            TimeSpan refreshRate)
+        {
+            _imdsAccessMethods = imdsAccessMethods;
+            _refreshRate = refreshRate;
+
             _credentialsRetrieverTimer = new Timer(RenewCredentials, null, TimeSpan.Zero, _neverTimespan); // This invokes synchronous calls in seperate thread.
             FeatureIdSources.Add(UserAgentFeatureId.CREDENTIALS_IMDS);
         }
 
         #region Overrides
-        
+
         /// <summary>
         /// Returns a copy of the most recent instance profile credentials.
         /// </summary>
         public override ImmutableCredentials GetCredentials()
         {
-            CheckIsIMDSEnabled();
-            ImmutableCredentials credentials = null;
+            // The logger instance is not cached as member variable because the static constructor
+            // can fire before the adaptor plug ins have been registered causing them to not
+            // send the logs to the correct destination.
+            var logger = Logger.GetLogger(typeof(DefaultInstanceProfileAWSCredentials));
+            _imdsAccessMethods.CheckIsIMDSEnabled();
+            ImmutableCredentials credentials;
+
+            // Copy current assignment of _lastRetrievedCredentials into a local variable in case another thread updates it.
+            var localLastRetrievedCredentials = _lastRetrievedCredentials;
 
-            // Try to acquire read lock. The thread would be blocked if another thread has write lock.
-            if (_credentialsLock.TryEnterReadLock(_credentialsLockTimeout))
+            if (null != localLastRetrievedCredentials && !localLastRetrievedCredentials.IsExpiredWithin(TimeSpan.Zero))
+            {
+                credentials = localLastRetrievedCredentials.Credentials;
+            }
+            else
             {
-                try
+                logger.DebugFormat("Waiting on lock to refresh ECS IMDS");
+                if (_credentialsSemaphore.Wait(_credentialsLockTimeout))
```

Comment:
Missing _isDisposed Check in GetCredentials Methods
csharp

Problem: If Dispose() is called while GetCredentials() or GetCredentialsAsync() is executing, they could attempt to wait on a disposed semaphore.

Impact: ObjectDisposedException could be thrown to callers instead of a more meaningful error.

Fix: Add early check: